src/client/dfuse/pil4dfs/aio.c

```diff
@@ -0,0 +1,506 @@
+		if (rc)
+			DL_ERROR(rc, "daos_event_fini() failed");
+		D_FREE(ev);
+		(*num_ev)++;
+	}
+	D_MUTEX_UNLOCK(&ctx->lock);
+}
+
+#define AIO_EQ_DEPTH MAX_EQ
+
+/* poll all EQs in our list */
+static void
+aio_poll_eqs(struct d_aio_ctx *ctx, long min_nr, long nr, struct io_event *events, int *num_ev)
+{
+	int                i, j;
+	int                rc, rc2;
+	struct daos_event *eps[AIO_EQ_DEPTH + 1] = {0};
+	struct d_aio_ev   *p_aio_ev;
+
+	ev_dequeue_batch(ctx, min_nr, nr, events, num_ev);
+	if (*num_ev >= min_nr)
+		return;
+
+	/* loop over all EQs */
+	for (i = 0; i < d_aio_eq_count_g; i++) {
+		if (atomic_load_relaxed(&aio_eq_list[i].n_op_queued) == 0)
+			continue;
+
+		rc = daos_eq_poll(aio_eq_list[i].eq, 0, DAOS_EQ_NOWAIT, AIO_EQ_DEPTH, eps);
+		if (rc < 0)
+			DL_ERROR(rc, "daos_eq_poll() failed");
+
+		for (j = 0; j < rc; j++) {
+			if (eps[j]->ev_error) {
+				DS_ERROR(eps[j]->ev_error, "daos_eq_poll() error");
+			} else {
+				atomic_fetch_add_relaxed(&aio_eq_list[i].n_op_queued, -1);
+				atomic_fetch_add_relaxed(&aio_eq_list[i].n_op_done, 1);
+				p_aio_ev = container_of(eps[j], struct d_aio_ev, ev);
+				if (p_aio_ev->ctx == ctx) {
+					/* append to event list */
+					D_MUTEX_LOCK(&ctx->lock);
+					events[*num_ev].obj = p_aio_ev->piocb;
+					events[*num_ev].res = p_aio_ev->piocb->u.c.nbytes;
+
+					rc2 = daos_event_fini(&p_aio_ev->ev);
+					if (rc2)
+						DL_ERROR(rc, "daos_event_fini() failed");
+					(*num_ev)++;
+					D_MUTEX_UNLOCK(&ctx->lock);
+					D_FREE(p_aio_ev);
+				} else {
+					/* need to append context's finished queue */
+					ev_enqueue(p_aio_ev->ctx, p_aio_ev);
+				}
+			}
+		}
+	}
+
+	/* try again after polling */
+	ev_dequeue_batch(ctx, min_nr, nr, events, num_ev);
+	if (*num_ev >= min_nr)
+		return;
+}
+
+static int
+io_getevents_sys(io_context_t ctx, long min_nr, long max_nr, struct io_event *events,
+		 struct timespec *timeout)
+{
+	return syscall(__NR_io_getevents, ctx, min_nr, max_nr, events, timeout);
+}
+
+int
+io_getevents(io_context_t ctx, long min_nr, long nr, struct io_event *events,
+	     struct timespec *timeout)
+{
+	d_aio_ctx_t    *aio_ctx_obj = (d_aio_ctx_t *)ctx;
+	io_context_t    ctx_real    = aio_ctx_obj->ctx;
+	int             op_done     = 0;
+	struct timespec times_0;
+	struct timespec times_1;
+	struct timespec dt;
+
+	if (next_io_getevents == NULL)
+		next_io_getevents = io_getevents_sys;
+
+	if (!d_hook_enabled)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+
+	if (aio_ctx_obj->depth == 0)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+	if (!aio_ctx_obj->on_dfs)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+	if (!aio_ctx_obj->inited) {
+		DS_ERROR(EINVAL, "event queue is not initialized yet");
+		return (-EINVAL);
+	}
+	if (timeout)
+		clock_gettime(CLOCK_REALTIME, &times_0);
+
+	if (min_nr > AIO_EQ_DEPTH)
+		min_nr = AIO_EQ_DEPTH;
+
+	while (1) {
+		aio_poll_eqs(aio_ctx_obj, min_nr, nr, events, &op_done);
```

Comment:
Now each aio context uses only one EQ. In case of the number of aio contexts is larger than the number of EQs for aio (limit is 16 currently), multiple aio contexts might share EQ.

Reply:
I think it would be better to just keep it 1 separate EQ for each context and not put a limitation on the number of EQs in pil4dfs for aio. Internally we already have a cap on the number of network contexts that are shared at some point between EQs, so this won't cause a resource limit.
let's address though in a follow on PR.

Reply:
Good to know. I will make such change in a follow up PR. Thank you!

Reply:
@mchaarawi Ticket was created for fixing this issue. https://daosio.atlassian.net/browse/DAOS-15803